File: web/src/api/backtest.ts
import axios from 'axios';

const api = axios.create({
  baseURL: '/api/v1',
  timeout: 30000,
});

export interface BacktestRequest {
  strategy: string;
  symbol: string;
  start_date: string;
  end_date: string;
  initial_cash: number;
  params: Record<string, any>;
}

export interface TradeMarker {
    date: string;
    type: 'buy' | 'sell';
    price: number;
    amount: number;
}

export interface BacktestResult {
  final_value: number;
  pnl: number;
  sharpe_ratio: number;
  max_drawdown: number;
  total_trades: number;
  win_rate: number;
  chart_data: Array<[string, number, number, number, number]>; // Date, Open, Close, Low, High
  trade_markers: TradeMarker[];
}

export interface TaskResponse {
    task_id: string;
    status: string;
}

export interface TaskStatusResponse {
    state: string;
    status?: string;
    result?: BacktestResult; // Only present if state is SUCCESS
    error?: string;
}

export const runBacktest = async (data: BacktestRequest) => {
  // Updated to return task_id
  const response = await api.post<TaskResponse>('/backtest/run', data);
  return response.data;
};

export const getBacktestStatus = async (taskId: string) => {
    const response = await api.get<TaskStatusResponse>(`/backtest/status/${taskId}`);
    return response.data;
}

export const getStrategies = async () => {
  const response = await api.get<string[]>('/backtest/strategies');
  return response.data;
};

File: api/routers/backtest.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Dict, Any
from celery.result import AsyncResult
from tasks.worker import run_backtest_celery

router = APIRouter()

class BacktestRequest(BaseModel):
    strategy: str
    symbol: str
    start_date: str
    end_date: str
    params: Dict[str, Any] = {}
    initial_cash: float = 10000.0

@router.post("/run")
async def run_backtest(request: BacktestRequest):
    """
    异步提交回测任务
    """
    task = run_backtest_celery.delay(
            strategy_name=request.strategy,
            symbol=request.symbol,
            params=request.params,
            start_date=request.start_date,
            end_date=request.end_date
        )
    return {"task_id": task.id, "status": "submitted"}

@router.get("/status/{task_id}")
async def get_backtest_status(task_id: str):
    """
    查询回测任务状态
    """
    task_result = AsyncResult(task_id)
    
    if task_result.state == 'PENDING':
        return {"state": "PENDING", "status": "Pending..."}
    elif task_result.state != 'FAILURE':
        result = task_result.result
        if result and isinstance(result, dict) and result.get("status") == "failed":
             return {
                "state": "FAILURE",
                "error": result.get("error")
            }
            
        return {
            "state": task_result.state,
            "result": result
        }
    else:
        return {
            "state": "FAILURE",
            "error": str(task_result.result)
        }

@router.get("/strategies")
def list_strategies():
    return ["SmaCross"]
